Frameworks/CoreGraphics/CGPath.mm

```diff
@@ -462,40 +413,53 @@ void CGPathAddPath(CGMutablePathRef path, const CGAffineTransform* transform, CG
+    path->SetStartingPoint(toAdd->GetStartingPoint());
 }
 
 /**
  @Status Interoperable
 */
 void CGPathAddEllipseInRect(CGMutablePathRef path, const CGAffineTransform* transform, CGRect rect) {
     RETURN_IF(!path);
+    CGRect containingRectangle = rect;
+    // First we must transform the coordinates of the rectangle if necessary.
+    if (transform) {
+        containingRectangle = CGRectApplyAffineTransform(rect, *transform);
+    }
 
-    // Determine the control point offset multiplier to create 4 arcs
-    CGFloat offsetMultiplier = _CGPathControlPointOffsetMultiplier(M_PI_2);
+    CGFloat radiusX = containingRectangle.size.width / 2.0;
+    CGFloat radiusY = containingRectangle.size.height / 2.0;
+    CGPoint center = CGPointMake(containingRectangle.origin.x + radiusX, containingRectangle.origin.y + radiusY);
 
-    CGFloat xControlPointOffset = offsetMultiplier * CGRectGetWidth(rect) / 2.0;
-    CGFloat yControlPointOffset = offsetMultiplier * CGRectGetHeight(rect) / 2.0;
+    D2D1_ELLIPSE ellipse = D2D1::Ellipse(_CGPointToD2D_F(center), radiusX, radiusY);
 
-    CGFloat minX = CGRectGetMinX(rect);
-    CGFloat midX = CGRectGetMidX(rect);
-    CGFloat maxX = CGRectGetMaxX(rect);
+    ComPtr<ID2D1Factory> factory;
+    FAIL_FAST_IF_FAILED(_CGGetD2DFactory(&factory));
+    ComPtr<ID2D1EllipseGeometry> ellipseGeometry;
 
-    CGFloat minY = CGRectGetMinY(rect);
-    CGFloat midY = CGRectGetMidY(rect);
-    CGFloat maxY = CGRectGetMaxY(rect);
+    FAIL_FAST_IF_FAILED(factory->CreateEllipseGeometry(&ellipse, &ellipseGeometry));
 
-    // Move to the center of the ellipse
-    CGPathMoveToPoint(path, transform, maxX, midY);
+    path->ClosePath();
 
-    // Add the 4 curves for the ellipse
-    CGPathAddCurveToPoint(path, transform, maxX, midY + yControlPointOffset, midX + xControlPointOffset, maxY, midX, maxY);
-    CGPathAddCurveToPoint(path, transform, midX - xControlPointOffset, maxY, minX, midY + yControlPointOffset, minX, midY);
-    CGPathAddCurveToPoint(path, transform, minX, midY - yControlPointOffset, midX - xControlPointOffset, minY, midX, minY);
-    CGPathAddCurveToPoint(path, transform, midX + xControlPointOffset, minY, maxX, midY - yControlPointOffset, maxX, midY);
+    // Create temp vars for new path/sink
+    ComPtr<ID2D1PathGeometry> newPath;
+    ComPtr<ID2D1GeometrySink> newSink;
 
-    CGPathCloseSubpath(path);
+    // Open a new path that the contents of the old path will be streamed into. We cannot re-use the same path as it is now closed
+    // and cannot be opened again. We use the newPath variable because the factory was returning the same pointer for some strange
+    // reason so this will force it to do otherwise.
+    FAIL_FAST_IF_FAILED(factory->CreatePathGeometry(&newPath));
+    FAIL_FAST_IF_FAILED(newPath->Open(&newSink));
+
+    FAIL_FAST_IF_FAILED(ellipseGeometry->CombineWithGeometry(path->GetPathGeometry().Get(), D2D1_COMBINE_MODE_XOR, NULL, newSink.Get()));
+
+    path->SetPathGeometry(newPath);
```

Comment:
>SetPathGeometry [](start = 10, length = 15)

There shouldn't be a world where you can set one but keep an old value in the other.

How about `SetPathGeometryAndSink(xxx, nullptr)`, `SetPathGeometryAndSink(xxx, yyy)`